kernel/EigenSolvers/JacobiDavidsonMaxwell.cpp

```diff
@@ -0,0 +1,889 @@
+
+    Vec small_evects[this->search_space_maxsize];
+    Vec tmp_v[this->search_space_maxsize];
+    Vec residue_vect_copy, correction_vect;
+    Vec q;
+    PetscInt k, idx_evect, ii;
+    PetscBool found, stop;
+    PetscScalar *vec_values;
+    PetscErrorCode ierr;
+
+    this->search_space_minsize = nev + 1;
+    this->nconverged = 0;
+
+    logger(INFO, "Sovling EigenValue problem using JacobiDavidsonSolver");
+
+    if (this->search_space_minsize >= search_space_maxsize) {
+        PetscPrintf(PETSC_COMM_WORLD, " Too many eigenvalues required");
+        exit(0);
+    }
+
+    // init the solver matrices and vectors
+    initializeMatrices();
+    initializeVectors();
+    initializeSearchSpace(nev);
+
+    // rayleigh quotient
+    computeRayleighQuotient(this->search_vect, &rho);
+
+    // compute initial residue vector
+    computeResidueVector(this->search_vect, rho, this->residue_vect);
+
+    // duplicate residue vector and create the correction vector
+    ierr = VecDuplicate(this->residue_vect, &residue_vect_copy);
+    CHKERRABORT(PETSC_COMM_WORLD, ierr);
+    ierr = VecDuplicate(this->residue_vect, &correction_vect);
+    CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+    k = 0;
+    stop = PETSC_FALSE;
+
+    for (this->iter = 0; this->iter < this->maxIter; this->iter++) {
+
+        logger(DEBUG, "JacobiDavidsonSolver iteration : %, k = %, rho = %",
+               iter, k, rho);
+
+        // determine eta
+        eps = (PetscReal)(std::rand()) / (PetscReal)(RAND_MAX);
+        this->eta = (this->iter > 0 && eps < 0.5) ? rho : this->tau;
+
+        // copy res to res_new and left project
+        ierr = VecCopy(this->residue_vect, residue_vect_copy);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+        computeProjection(residue_vect_copy, this->Qt);
+
+        // solve correction equation
+        solveCorrectionEquation(residue_vect_copy, correction_vect);
+
+        // ortho correction vector
+        computeProjection(correction_vect, this->Qt);
+
+        // project correction vector
+        projectCorrectionVector(correction_vect);
+
+        // create search vector
+        ierr = VecCopy(correction_vect, search_vect);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+        computeProjection(search_vect, this->V);
+        normalizeVector(search_vect);
+
+        k++;
+
+        // Add column to V
+        ierr = BVInsertVec(this->V, this->V_current_size, search_vect);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+        this->V_current_size++;
+        ierr = BVSetActiveColumns(this->V, 0, this->V_current_size);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+        // compute eigenpairs on small subspace
+        ierr = computeSmallEigenvalues(small_evals, small_evects);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+        // create q
+        ierr = VecDuplicate(this->residue_vect, &q);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+        found = PETSC_TRUE;
+        idx_evect = 0;
+        while (found) {
```

Comment:
Stability of the small eigenvectors (and eigenvalues) between processes will become quite important here. If they are not stable enough one process may accept the eigenvalues, while others might not.